ch5/ex-5-17-qsort-fields.c

```diff
@@ -0,0 +1,228 @@
+/* Exercise 5-17
+ *
+ * Add a field-handling capability, so sorting may be done on fields within
+ * lines, each field sorted according to an independent set of options.  (The
+ * index for this book was sorted with -df for the index category and -n for the
+ * page numbers.)
+ */
+
+#include <ctype.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include "lines.h"
+#include "quicksort.h"
+
+#define CHARCOMP int (*)(char *, char *)
+#define VOIDCOMP int (*)(void *, void *)
+
+void parseargs(int argc, char *argv[]);
+char **getsortdata(char *line);
+int compare(char **s1, char **s2);
+
+char *line(char *sortdata[]);
+char *fieldspace(char *sortdata[]);
+void fold(char *value);
+void dirsort(char *value);
+int strcmpr(char *, char *);
+int numcmp(char *, char *);
+int numcmpr(char *, char *);
+void freestuff(char ***linedata, int nlines);
+
+int *sortidxs; /* indexes of fields to sort by, ordered by priority */
+int nsortflds; /* number of fields we're sorting by */
+int maxsortidx; /* highest index of the fields we're sorting by */
+int (**compares)(char *, char *); /* comparison function for each sort field */
+int *folds; /* whether to fold each sort field */
+int *dirsorts; /* whether to use dir sort for each sort field */
+
+int main(int argc, char *argv[])
+{
+	int i, nlines;
+	char *buff, **lines, ***linedata;
+
+	parseargs(argc, argv);
+
+	if ((nlines = readlines(&buff, &lines)) == LNS_ERROR) {
+		printf("input too big to sort\n");
+		return 0;
+	}
+
+	linedata = malloc(nlines * sizeof(char **));
+	for (i = 0; i < nlines; i++)
+		linedata[i] = getsortdata(lines[i]);
+
+	quicksort((void **)linedata, 0, nlines - 1, (VOIDCOMP)compare);
+
+	for (i = 0; i < nlines; i++)
+		lines[i] = line(linedata[i]);
+
+	writelines(lines, nlines);
+
+	freestuff(linedata, nlines);
+	freelines(buff, lines);
+	return 0;
+}
+
+int compare(char **sortdata1, char **sortdata2)
+{
+	int i, rslt = 0;
+
+	for (i = 0; i < nsortflds && rslt == 0; i++)
+		rslt = (*compares[i])(sortdata1[i], sortdata2[i]);
+	return rslt;
+}
+
+void parseargs(int argc, char *argv[])
+{
+	int i;
+	int maxsortcount = argc - 1;
```

Comment:
What would happen if `argc` is `0`?

Reply:
Oh it'll break :-) 

Tail is the only exercise so far that talked about handling 'unreasonable' input.   Without that requirement I'm back in cowboy mode.  I did start writing a check for `0` but then thought either check input properly or not at all, and doing it properly for this execice would be a lot of work, there are lot of ways to break it.